Unregister Block Example (#138)
10up/gutenberg-best-practices: reference/03-Blocks/unregister-block.md

`block-variations/unregister-core-embed-variation.js`
```js
import domReady from '@wordpress/dom-ready';
import { getBlockVariations, unregisterBlockVariation } from '@wordpress/blocks';

const allowedEmbedBlocks = ['vimeo', 'youtube'];

domReady(() => {
	const embedVariations = getBlockVariations('core/embed');
	embedVariations.forEach(({name}) => {
		if (allowedEmbedBlocks.includes(name)) {
			unregisterBlockVariation('core/embed', name);
		}
	});
});
```

## Allow List:
Opposite to the approach of `unregisterBlockType` would be to create a PHP allow list. In this scenario a developer would curate the specific blocks that they want to have shown to an editor. In these situations it is advised to always include common utility blocks such as: `core/block`, `core/missing`, `core/pattern`, `core/template-part` to prevent issues with the block editor. 

> **NOTE** - When using an allow list it is advised to review WordPress releases for any changes in existing blocks. As an example WordPress 6.1 refactored the list block from one singular block to an inner blocks structure with a list item. Anyone that used an allow list at that point needed to go in and update it to include the core/list-item block.

Reviewer: Docusaurus has a special syntax for these notes: https://docusaurus.io/docs/markdown-features/admonitions

```suggestion
:::caution
When using an allow list it is advised to review WordPress releases for any changes in existing blocks. As an example WordPress 6.1 refactored the list block from one singular block to an inner blocks structure with a list item. Anyone that used an allow list at that point needed to go in and update it to include the core/list-item block.
::::
```

---
sidebar_position: 10
---

# Unregister a Block

## Reference Guide WordPress Documentation:
* [unregisterBlockType](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-blocks/#unregisterblocktype)

## Use Case:

> Gutenberg comes with a lot of core blocks. When creating custom builds for clients all of these core blocks are generally not needed nor should they be used by the client. In these situations it is necessary to disable the blocks from the block editor so that a user can not access them when creating new pages and posts with the block editor. This can be done using the `unregisterBlockType` function and passing in the blocks which should be disallowed within a `.js` file. 

Reviewer: Whilst I agree with this statement here I would reword it a little bit to put emphasis on the fact that we are removing some of these blocks in order to remove cognitive load/confusion from editors.

## Code Example:
This example showcases how an array of blocks can be passed into a single function and looped through to disable multiple blocks at the same time. 

`block-filters/unregister-blocks.js`

> ```bash

Reviewer: ```suggestion
```js
```

> const { unregisterBlockType } = wp.blocks;

Reviewer: We tend to use the import syntax in the documentation here as it is standard on all modern builds :) 

```suggestion
import { unregisterBlockType } from '@wordpress/blocks';
```

const disallowedBlocks = [
	'core/preformatted',
	'core/table',
	'core/code',
	'core/video',
	'core/verse',
	'core/avatar',
	'core/audio',
	'core/calendar',
	'core/freeform',
];

> wp.domReady(() => {

Reviewer: again this should be listed in the imports at the top and then only call the `domReady` function here.

> 	wp.blocks.getBlockTypes().forEach(function (blockType) {

Reviewer: Same here :)

> 	wp.blocks.getBlockTypes().forEach(function (blockType) {
> 		if (disallowedBlocks.indexOf(blockType.name) !== -1) unregisterBlockType(blockType.name);
> 	});

Reviewer: I reworked the code example slightly to be a little easier to parse. Also as per our coding standards we tend to avoid inline control statements because they can easily cause issues when refactoring :) 

So 

```js
// instead of this
if ( condition ) doSomething();

// we always want to write out the full if statement
if ( condition ) {
    doSomething();
}
```
 
```suggestion
	const blocks = getBlockTypes();
	blocks.forEach(({ name }) {
		if (disallowedBlocks.includes(name)) {
			unregisterBlockType(name);
		}
	});
```

});
```

### Another Example - Variations:

> In this use case only `core/embed` has many variations of embeds available to a user but per client requirements only Vimeo and YouTube should be available as other platforms are not supported. In this use case `unregisterBlockVariation` is used since it is related to a `variation` of a block.

Reviewer: ```suggestion
Something that also comes up all the time is the need to remove individual block variations. The core embed block for example uses block variations to expose the various embed providers as separate items in the block inserted. If a client however has strict requirements to only ever embed videos using YouTube for example, we want to hide all the other embed providers. We can achieve that using the `unregisterBlockVariation` function.
```

`block-variations/unregister-core-embed-variation.js`

> ```bash

Reviewer: ```suggestion
```js
```

> const allowedEmbedBlocks = ['vimeo', 'youtube'];
> 
> wp.domReady(function () {
> 	wp.blocks.getBlockVariations('core/embed').forEach(function (blockVariation) {
> 		if (allowedEmbedBlocks.indexOf(blockVariation.name) === -1) {
> 			wp.blocks.unregisterBlockVariation('core/embed', blockVariation.name);
> 		}
> 	});
> });

Reviewer: ```suggestion
import domReady from '@wordpress/dom-ready';
import { getBlockVariations, unregisterBlockVariation } from '@wordpress/blocks';

const allowedEmbedBlocks = ['vimeo', 'youtube'];

domReady(() => {
	const embedVariations = getBlockVariations('core/embed');
	embedVariations.forEach(({name}) => {
		if (allowedEmbedBlocks.includes(name)) {
			unregisterBlockVariation('core/embed', name);
		}
	});
});
```
